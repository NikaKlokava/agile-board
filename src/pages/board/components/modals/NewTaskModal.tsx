import { ModalWrapper } from "../../../../shared/components/modal_wrapper";
import { Button } from "../../../../shared/components/button";
import { FieldArray, Form, Formik } from "formik";
import { Input } from "../../../../shared/components/input";
import { Select } from "../../../../shared/components/select";
import { initialTaskData, TaskSchema } from "../../../../utils/utils";
import { FieldName } from "../../../../shared/components/field_name";
import { DescriptionField } from "../../../../shared/components/description/DescriptionField";
import { AddBtn } from "../../../../shared/components/add_button";
import { v4 as uuidv4 } from "uuid";
import { addNewTask } from "../../../../redux/reducers/tasksSlice";
import { useAppDispatch, useAppSelector } from "../../../../redux/hooks/hook";
import cl from "./modal_styles.module.css";
import { saveTaskData } from "../../../../redux/reducers/tasksSlice";

type Props = {
  onClose: () => void;
};

export const NewTaskModal = ({ onClose }: Props) => {
  const activeBoard = useAppSelector((state) => state.activeBoard);
  const dispatch = useAppDispatch();

  const initialData = {
    ...initialTaskData,
    columnTitle: activeBoard.columns[0].title,
  };

  const onSubmit = (values: AddNewTaskAction) => {
    const columnIndex = activeBoard.columns.findIndex(
      (column) => column.title === values.columnTitle
    );
    const columnUuid = activeBoard.columns[columnIndex].uuid;

    const newTask: Task = {
      uuid: uuidv4(),
      boardUuid: activeBoard.uuid,
      columnUuid,
      title: values.title,
      description: values.description,
      subtasks: values.subtasks,
      time: new Date().getTime(),
    };
    dispatch(addNewTask(newTask));
    dispatch(saveTaskData({ newTask: newTask }));
    onClose();
  };

  return (
    <ModalWrapper onWrapperClick={onClose}>
      <Formik
        initialValues={initialData}
        validationSchema={TaskSchema}
        onSubmit={(values) => {
          const subtasks = values.subtasks
            ?.filter((subtask) => subtask && subtask.trimStart().length !== 0)
            ?.map((e) => ({
              text: e.trimStart(),
              uuid: uuidv4(),
              checked: false,
            }));

          onSubmit({
            ...values,
            subtasks: subtasks,
          });
        }}
      >
        {(props) => (
          <Form
            onSubmit={(e) => {
              e.preventDefault();
              props.handleSubmit();
            }}
            className={cl.form_container}
          >
            <h2 className={cl.modal_title} data-testid="new-task-modal">
              Add New Task
            </h2>
            <FieldName formikName="title" fieldName={"Task name"} />
            {props.errors.title && props.touched.title && (
              <p style={{ color: "red" }}>{props.errors.title}</p>
            )}
            <DescriptionField />
            <FieldArray
              name="subtasks"
              render={(arrayHelpers) => (
                <>
                  <div className={cl.container}>
                    <p className={cl.title}>Subtasks</p>
                    {props.values.subtasks?.map((_, index) => (
                      <Input
                        key={index}
                        formikName={`subtasks[${index}]`}
                        remove={arrayHelpers.remove}
                        index={index}
                      />
                    ))}
                    <AddBtn
                      add={arrayHelpers.push}
                      text="Add New Subtask"
                      testid={"add-new-subtask-btn"}
                    />
                  </div>
                </>
              )}
            />
            <Select />
            <Button
              text="Create Task"
              withIcon={false}
              newClass="center"
              testid={"create-task-btn"}
              type="submit"
            />
          </Form>
        )}
      </Formik>
    </ModalWrapper>
  );
};
