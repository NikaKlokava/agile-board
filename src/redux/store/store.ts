import { legacy_createStore } from "redux";
import { boardsReducer } from "../reducers/boardReducer";

const store = legacy_createStore(boardsReducer);

export default store;

// export type RootState = ReturnType<typeof store.getState>

// export interface RootState {
//     user: UserState;
//     loader: LoaderState;
//   }
// export type AppDispatch = typeof store.dispatch
