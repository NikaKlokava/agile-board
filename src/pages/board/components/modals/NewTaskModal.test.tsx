import { Provider } from "react-redux";
import renderer from "react-test-renderer";
import store from "../../../../redux/store/store";
import { NewTaskModal } from "./NewTaskModal";

describe("Test the NewTaskModal component", () => {
  test("The NewTaskModal renders correctly", () => {
    const newTaskModalSnap = renderer
      .create(
        <Provider store={store}>
          <NewTaskModal onClose={() => {}} />
        </Provider>
      )
      .toJSON();
    expect(newTaskModalSnap).toMatchSnapshot();
  });
});
