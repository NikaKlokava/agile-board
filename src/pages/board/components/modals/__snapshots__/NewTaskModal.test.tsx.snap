// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the NewTaskModal component The NewTaskModal renders correctly 1`] = `
Array [
  <div
    className="modal_wrapper"
  />,
  <div
    className="modal_container"
  >
     
    <h2
      className="modal_title"
      data-testid="new-task-modal"
    >
      Add New Task
    </h2>
    <div
      className="container"
    >
      <p
        className="title"
      >
        Task name
      </p>
      <input
        autoComplete="off"
        className="input_style"
        name="title"
        onChange={[Function]}
        placeholder="e.g Take coffee break"
        spellCheck={false}
        type="text"
      />
    </div>
    <div
      className="container"
    >
      <p
        className="title"
      >
        Description
      </p>
      <textarea
        className="textarea"
        name="description"
        onChange={[Function]}
        placeholder="e.g. It's always good to take a break..."
        spellCheck={false}
      />
    </div>
    <div
      className="container style_container"
    >
      <p
        className="title"
      >
        Subtasks
      </p>
      <div
        className="input_container"
      >
        <input
          autoComplete="off"
          className="input_style"
          name="subtasks[0].text"
          onChange={[Function]}
          placeholder="e.g Take coffee break"
          spellCheck={false}
          type="text"
        />
        <div
          className="delete_icon"
          onClick={[Function]}
        />
      </div>
      <div
        className="input_container"
      >
        <input
          autoComplete="off"
          className="input_style"
          name="subtasks[1].text"
          onChange={[Function]}
          placeholder="e.g Take coffee break"
          spellCheck={false}
          type="text"
        />
        <div
          className="delete_icon"
          onClick={[Function]}
        />
      </div>
      <div
        className="button"
        data-testid="add-new-subtask-btn"
        onClick={[Function]}
      >
        <div
          className="button_icon"
        />
        <div
          className="button_text"
        >
          Add New Subtask
        </div>
      </div>
    </div>
    <div
      className="container"
    >
      <p
        className="title"
      >
        Current Status
      </p>
      <select
        name="columnTitle"
        onChange={[Function]}
      />
    </div>
    <div
      className="button center"
      data-testid="create-task-btn"
      onClick={[Function]}
    >
      <div
        className="button_text"
      >
        Create Task
      </div>
    </div>
  </div>,
]
`;
